Add code for generating the Implementation Report (#72)
w3c/rdf-canon: reports/README.md

This is a collection of individual
[EARL reports](https://www.w3.org/TR/EARL10-Schema/) for
test subjects claiming JSON-LD processor conformance.

The consolidated report is saved to `index.html` generated
using the
[`earl-report` Ruby gem](https://rubygems.org/gems/earl-report).
Run it as follows:

```sh
$ gem install earl-report
$ rm -f manifests.nt && rake manifests.nt
$ earl-report --format json -o earl.jsonld *.ttl
$ earl-report --json --format ttl -o earl.ttl earl.jsonld
$ earl-report --json --format html --template template.haml -o index.html earl.jsonld
```

## Instructions for submitting implementation reports

Reports should be submitted in Turtle format to
  <a href="mailto:public-rch-wg@w3.org">Public RCH WG</a> or via a Pull
  Request to the [w3c/rdf-canon repository](https://github.com/w3c/rdf-canon/pulls).

Tests should be run using the test manifests defined in the
  [Test Manifests](#test-manifests) Section.

Include an `earl:Assertion` for each test, referencing the test
  resource from the associated manifest and the test subject being
  reported upon. See the example below:

> ```

Reviewer: ~~~suggestion
```turtle
~~~

[ a               earl:Assertion ;
  earl:assertedBy <--your-developer-identifier--> ;
  earl:subject    <--your-software-identifier--> ;
  earl:test       <--uri-of-test-from-manifest> ;
  earl:result     [ a            earl:TestResult ;
                    earl:outcome earl:passed ;
                    dc:date      "2023-01-25T10:18:04-08:00"^^xsd:dateTime 
                  ] ;
    earl:mode     earl:automatic 
] .
```

The Test Subject should be defined as a `doap:Project`, including the name,
  homepage, and developer(s) of the software
  (see [DOAP](https://github.com/edumbill/doap/wiki)),
  and optionally including the
  project description and programming language. An example test subject description is the following:

> ```

Reviewer: ~~~suggestion
```turtle
~~~

<> foaf:primaryTopic        <--your-software-identifier--> ;
   dc:issued                "2016-12-26T10:18:04-08:00"^^xsd:dateTime ;
   foaf:maker               <--your-developer-identifier--> .

<--your-software-identifier--> 
   a                        doap:Project , 
                            earl:TestSubject , 
                            earl:Software ;
  doap:name                 "My Cool RDF Canonicalizer" ;
  doap:release              [ doap:name     "--short name wih version number--" ;
                              doap:revision "--Software version number--" ;
                              doap:created  "2020-02-19"^^xsd:date ;
                            ] ;
  doap:developer            <--your-developer-identifier--> ;
  doap:homepage             <--your-software-homepace--> ;
  doap:description          "--your-project-description--"@en ;
  doap:programming-language "--your-implementation-language--" .
```

The software developer, either an organization or one or more individuals SHOULD be
  referenced from <code>doap:developer</code> using <a href="http://xmlns.com/foaf/spec">FOAF</a>. For example:</p>

> ```

Reviewer: ~~~suggestion
```turtle
~~~

This is a collection of individual
[EARL reports](https://www.w3.org/TR/EARL10-Schema/) for
test subjects claiming JSON-LD processor conformance.

The consolidated report is saved to `index.html` generated
using the

> [earl-report Ruby gem](https://rubygems.org/gems/earl-report).

Reviewer: ```suggestion
[`earl-report` Ruby gem](https://rubygems.org/gems/earl-report).
```

```

## Instructions for submitting implementation reports

Reports should be submitted in Turtle format to

>   <a href="mailto:public-rch-wg@w3.org">Public RCH WG</a> or via a Pull

Reviewer: ```suggestion
Reports should be submitted in Turtle format via email to
    [Public RCH WG](mailto:public-rch-wg@w3.org), or via a Pull

```

>   Request to the <a href="https://github.com/w3c/rdf-canon/pulls">w3c/rdf-canon</a>.

Reviewer: ```suggestion
  Request to the [w3c/rdf-canon repository](https://github.com/w3c/rdf-canon/pulls).
```

Author: I'm fine with this, but since it's largely copied from the Haml (HTML) file, and HTML is compatible with Markdown, I didn't feel the need. But, when reading the file as Markdown, it is easier for a human to parse.

> Tests should be run using the test manifests defined in the
>   <a href="#test-manifests">Test Manifests</a> Section.

Reviewer: ```suggestion
Tests should be run using the test manifests defined in the
  [Test Manifests](#test-manifests) Section.
```

> Include an <code>earl:Assertion</code> for each test, referencing the test

Reviewer: ```suggestion
Include an `earl:Assertion` for each test, referencing the test
```

  resource from the associated manifest and the test subject being
  reported upon. See the example below:

```
[ a earl:Assertion;
    earl:assertedBy &lt;--your-developer-identifier--&gt;
    earl:subject <--your-software-identifier-->;
    earl:test <--uri-of-test-from-manifest>;
    earl:result [
      a earl:TestResult;
        earl:outcome earl:passed;
        dc:date "2023-01-25T10:18:04-08:00"^^xsd:dateTime];
    earl:mode earl:automatic ] .

> ```

Reviewer: ~~~suggestion
```
[ a               earl:Assertion ;
  earl:assertedBy <--your-developer-identifier--> ;
  earl:subject    <--your-software-identifier--> ;
  earl:test       <--uri-of-test-from-manifest> ;
  earl:result     [ a            earl:TestResult ;
                    earl:outcome earl:passed ;
                    dc:date      "2023-01-25T10:18:04-08:00"^^xsd:dateTime 
                  ] ;
    earl:mode     earl:automatic 
] .
```
~~~

> The Test Subject should be defined as a <code>doap:Project</code>, including the name,
>   homepage and developer(s) of the software
>   (see <a href="https://github.com/edumbill/doap/wiki">DOAP</a>).
>   Optionally, including the
>   project description and programming language. An example test subject description is the following:

Reviewer: ```suggestion
The Test Subject should be defined as a `doap:Project`, including the name,
  homepage, and developer(s) of the software
  (see [DOAP](https://github.com/edumbill/doap/wiki)),
  and optionally including the
  project description and programming language. An example test subject description is the following:
```

```
<> foaf:primaryTopic <--your-software-identifier--> ;
  dc:issued "2016-12-26T10:18:04-08:00"^^xsd:dateTime ;
  foaf:maker <--your-developer-identifier--> .

<--your-software-identifier--> a doap:Project, earl:TestSubject, earl:Software ;
  doap:name          "My Cool RDF Canonicalizer" ;
  doap:release [
    doap:name     "--short name wih version number--";
    doap:revision "--Software version number--" ;
    doap:created  "2020-02-19"^^xsd:date;
  ] ;
  doap:developer     <--your-developer-identifier--> ;
  doap:homepage      <--your-software-homepace--> ;
  doap:description   "--your-project-description--"@en ;
  doap:programming-language "--your-implementation-language--" .

> ```

Reviewer: ~~~suggestion
```
<> foaf:primaryTopic        <--your-software-identifier--> ;
   dc:issued                "2016-12-26T10:18:04-08:00"^^xsd:dateTime ;
   foaf:maker               <--your-developer-identifier--> .

<--your-software-identifier--> 
   a                        doap:Project , 
                            earl:TestSubject , 
                            earl:Software ;
  doap:name                 "My Cool RDF Canonicalizer" ;
  doap:release              [ doap:name     "--short name wih version number--" ;
                              doap:revision "--Software version number--" ;
                              doap:created  "2020-02-19"^^xsd:date ;
                            ] ;
  doap:developer            <--your-developer-identifier--> ;
  doap:homepage             <--your-software-homepace--> ;
  doap:description          "--your-project-description--"@en ;
  doap:programming-language "--your-implementation-language--" .
```
~~~

The software developer, either an organization or one or more individuals SHOULD be

>   referenced from <code>doap:developer</code> using <a href="http://xmlns.com/foaf/spec">FOAF</a>. For example:</p>

Reviewer: ```suggestion
The software developer(s), whether an organization or one or more individuals, SHOULD be
  referenced from `doap:developer` using [FOAF](http://xmlns.com/foaf/spec). For example:</p>
```

```
<--your-developer-identifier--> a foaf:Person, earl:Assertor;
  foaf:name "--My Name--";
  foaf:homepage <--my homepage--> .

> ```

Reviewer: ~~~suggestion
```
<--your-developer-identifier--> 
   a             foaf:Person , 
                 earl:Assertor ;
   foaf:name     "--My Name--" ;
   foaf:homepage <--my homepage--> .
```
~~~

> * `Rakefile` – Builds `manifest.ttl` from the test suite. (could also add tasks to build `earl.jsonld`, `earl.ttl`, and `index.html`)

Reviewer: ```suggestion
* `Rakefile` – Builds `manifest.ttl` from the test suite. (Tasks could also be added to build `earl.jsonld`, `earl.ttl`, and/or `index.html`)
```

> Running the rake task and earl-report requires the installation of Ruby, and the following gems:
> 
> * 'rdf-turtle'
> * 'json-ld'
> * 'haml'
> * 'kramdown'

Reviewer: ```suggestion
Running the `rake` task and `earl-report` requires installation of Ruby, and the following gems:

* `rdf-turtle`
* `json-ld`
* `haml`
* `kramdown`
```

## Generated files:

* `earl.jsonld` – JSON-LD representation of the consolidated EARL submissions build by earl-report as described above.
* `earl.ttl` – Turtle version of consolidated report, build from `earl.jsonld` by earl-report, as described above.
* `index.html` – The processed implementation report, built by earl-report using `earl.jsonld`, as described above.

> * `manifest.nt` – Local cache of the test-suite manifest files, updated via `rake manifest.nt`.

Reviewer: ```suggestion
* `earl.jsonld` — JSON-LD representation of the consolidated EARL submissions, built by `earl-report`, as described above.
* `earl.ttl` — Turtle version of consolidated report, build from `earl.jsonld` by `earl-report`, as described above.
* `index.html` — The processed implementation report, built by `earl-report` using `earl.jsonld`, as described above.
* `manifest.nt` — Local cache of the `test-suite` manifest files, updated via `rake manifest.nt`.
```

## Submission files:

> Any files (other than `earl.ttl`) ending in `.ttl` are taken as individual EARL reports of test-suite conformance. For example `ruby-earl.ttl` is the report for [Ruby RDF::Normalize](https://github.com/ruby-rdf/rdf-normalize).

Reviewer: ```suggestion
Any files ending in `.ttl` (other than `earl.ttl`) are taken as individual EARL reports of `test-suite` conformance. For example, `ruby-earl.ttl` is the report for [Ruby `RDF::Normalize`](https://github.com/ruby-rdf/rdf-normalize).
```
